### PR TITLE
Building a Rock Paper Scissors Game Using Socket

### PR DESCRIPTION
**Pull request by HarishTeens**
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      

#### Grammar Checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical Checks
- [ ] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution Guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

### CODE REVIEW

**Comment by HarishTeens** [comment]
This fixed #593

**Comment by hectorkambow** [comment]
👋 Thank you for submitting this PR - we will review it ASAP - all PRs go into our queue will be looked over before being approved to be published.

**Comment by hectorkambow** [comment]
Just an FYI - moving forward - lets please only create a PR once the .md file is fully ready to be reviewed - any changes can be made on your personal forked repo.
👍

**Comment by HarishTeens** [comment]
@hectorkambow It was a last minute idea to improve the article. Sorry about that, won't happen again 😁.

**Comment by louisefindlay23** [comment]
@hectorkambow, has this been pre-review checked?

**Comment by hectorkambow** [comment]
> @hectorkambow, has this been pre-review checked?

👋  @louisefindlay23 No - I don't believe it has been (through any sort of review yet). I will run a plagiarism check right now though.

**Comment by hectorkambow** [comment]
Plagiarism check done

File: articles/rock-paper-scissor-online/index.md [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+Here goes the file wise description of the Starter code,
+
+- ```public/index.html``` : A Home Page for the app with all the required UI Components. 
+- ```public/style.css``` : Styles for the entire application.
+- ```public/game.js``` : Client-Side JavaScript file to handle events.
+- ```app.js``` : Server Side NodeJS file for setting up the server.
+
+Note that the app is not functional yet, we need to write the necessary logic to make the app functional. Also, feel free to tweak the boiler code to match your style.
+
+There are two important files to focus on. One is  ```app.js``` that handles server-side logic. And the other is ```public/game.js``` that handles client-side logic.
```
Comment (louisefindlay23) [inline articles/rock-paper-scissor-online/index.md:36]: You only need one backtick for inline code elements. 

Join the next two sentences together. **logic and the other is**

File: articles/rock-paper-scissor-online/index.md [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+
+### Game Flow 
+
+![image](/engineering-education/nginx-reverse-proxy/rps.png)
+
+This is a Sequence diagram that shows the timeline of events. It's important to understand the flow before we start coding. So make sure you understand the diagram completely.
```
Comment (louisefindlay23) [inline articles/rock-paper-scissor-online/index.md:42]: coding so make sure **Join sentences**

File: articles/rock-paper-scissor-online/index.md [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+
+Read through the below points if you are not familiar with Sequence Diagrams.
+
+- Time flows downward.
+- Arrows represent events. The start of Arrow denotes the Emitter, end of arrow denotes the listener. For example, Player1 emits the `Create Game` Event, Socket Server listens to this event.
```
Comment (louisefindlay23) [inline articles/rock-paper-scissor-online/index.md:47]: Join your sentences

**of an** arrow, **the** end

Event **and the** Socket server listens

File: articles/rock-paper-scissor-online/index.md [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+- A rectangular box denotes waiting/processing time.
+
+Player1 emits the create Game Event. The server acknowledges and responds back with a room ID. Once player2 emits join Game Event with the same room ID. The server detects that both players have joined the room and shares the other player's info with each other. Once both of the players have their choice, the server processes the result and sends back to both players.
```
Comment (louisefindlay23) [inline articles/rock-paper-scissor-online/index.md:50]: Add linking words

**First,** Player1 emits the create Game Event **which t**he server **then** acknowledges and responds back with a room ID. Once**,** player2 emits join Game Event with the same room ID**, the** server **then** detects that both players have joined the room and shares the other player's info with each other. **Finally, o**nce both of the players have made **their** choice, the server processes the result and sends **it** back to both players.

File: articles/rock-paper-scissor-online/index.md [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+
+Now that the flow is clear, let's jump right into coding.
+
+### Create Game
+
+Let us start by writing logic for the Create Game Event in the `public/game.js` file.
```
Comment (louisefindlay23) [inline articles/rock-paper-scissor-online/index.md:56]: Let's **suits tone**

File: articles/rock-paper-scissor-online/index.md [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+$(".createBtn").click(function(){
+    firstPlayer=true;
+    const playerName=$("input[name=p1name").val();
+    socket.emit('createGame',{name:playerName});
+})
+```
+
+The above code listens for a click event on the Create Button. Once that event is triggered, the client grabs player1's name and emits a socket event named `createGame` . The variable `firstPlayer` identifies the player who started the game.  The use of this variable will be explained later.
+
+Now we write server-side code on `app.js` so that the server listens to this event.
```
Comment (louisefindlay23) [inline articles/rock-paper-scissor-online/index.md:69]: **Next,** we write **the** server-side code **in** `app.js` **Flow edit**

File: articles/rock-paper-scissor-online/index.md [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+        const roomID=randomstring.generate({length: 4});       
+        socket.join(roomID);        
+        players[roomID]=data.name;
+        socket.emit("newGame",{roomID:roomID});
+    })
+
+})
+```
+
+All event listeners/emitters would go inside the `io.on` block as mentioned above. `connection` is the default event listener provided by Socket.io. A `connection` event is emitted under the hood everytime a connection is established.
```
Comment (louisefindlay23) [inline articles/rock-paper-scissor-online/index.md:86]: Remove would

Socket.io and a connect event **- join sentences**

**every time**

File: articles/rock-paper-scissor-online/index.md [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+        socket.join(roomID);        
+        players[roomID]=data.name;
+        socket.emit("newGame",{roomID:roomID});
+    })
+
+})
+```
+
+All event listeners/emitters would go inside the `io.on` block as mentioned above. `connection` is the default event listener provided by Socket.io. A `connection` event is emitted under the hood everytime a connection is established.  
+The `CreateGame` listener creates a new Room with a random Room ID and adds the client to that room. Later the server emits an event `newGame` which contains the roomID for the created room.
```
Comment (louisefindlay23) [inline articles/rock-paper-scissor-online/index.md:87]: Added a comma after Later

File: articles/rock-paper-scissor-online/index.md [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+//New Game Created Listener
+socket.on("newGame",(data)=>{
+    $(".newRoom").hide(); 
+    $(".joinRoom").hide();
+    $("#message").html("Waiting for player 2, room ID is "+data.roomID).show();
+    roomID=data.roomID;
+})
+```
+
+The above code is added to `public/game.js`. This snippet lets player1 know that a room was created and player2 could use this roomID to join the room. Then it hides the Room elements and displays a message with Room ID.
```
Comment (louisefindlay23) [inline articles/rock-paper-scissor-online/index.md:99]: Added a comma after Then

**Review by louisefindlay23 — CHANGES_REQUESTED** [review]
@HarishTeens, I've highlighted some minor grammatical errors and suggested some improvements to aid readability and flow in the first half of your article. 

Please use my comments to look through the rest of your article. Consider joining shorted related sentences together with **and** to improve readability and use words like **First, Next and Finally** when providing a series of instructions to aid flow.

**Comment by HarishTeens** [comment]
> @HarishTeens, I've highlighted some minor grammatical errors and suggested some improvements to aid readability and flow in the first half of your article.
> 
> Please use my comments to look through the rest of your article. Consider joining shorted related sentences together with **and** to improve readability and use words like **First, Next and Finally** when providing a series of instructions to aid flow.

Thanks for the suggestions @louisefindlay23, I will make the changes as soon as possible

File: articles/rock-paper-scissor-online/index.md [hunk 10]
```diff
@@ -0,0 +1,260 @@
+A variety of applications use Socket on the modern web. For example chat applications, real-time databases and online multiplayer games. All use Socket under the hood. 
+
+### Introduction
+This article focuses on building a web application which lets you play Rock Paper Scissor online with your friends. A player has to create a room and share the unique Room ID with the other player. The other player can join the same room with this unique Room ID. Once both players have joined the room, they could start playing!
+
+There are two parts to this application, 
+- Front-end: To keep things simple, the front-end is built on basic HTML, CSS, JS with jQuery.
```
[the remark this answers is not in the document]
Comment (HarishTeens) [reply articles/rock-paper-scissor-online/index.md:7]: Got it 👍

File: articles/rock-paper-scissor-online/index.md [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+- Back-end: An Express server that uses Socket.io.
+
+### Prerequisites
+ 
+- HTML is the standard markup language for web page. Here's a useful [guide](https://developer.mozilla.org/en-US/docs/Learn/HTML/Introduction_to_HTML) for beginners.
+- CSS is style sheet language for improving the presentation of web pages. To learn more, check this [article](https://www.section.io/engineering-education/what-is-css/)
+- [BootStrap](https://getbootstrap.com/) is a popular HTML, CSS, JS library. An [article](https://getbootstrap.com/docs/4.0/getting-started/introduction/)  to get started.
```
[the remark this answers is not in the document]
Comment (HarishTeens) [reply articles/rock-paper-scissor-online/index.md:14]: Got it 👍

File: articles/rock-paper-scissor-online/index.md [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+- [jQuery](https://jquery.com/) is a fast and feature-rich Javascript library.
```
[the remark this answers is not in the document]
Comment (HarishTeens) [reply articles/rock-paper-scissor-online/index.md:15]: Got it 👍

File: articles/rock-paper-scissor-online/index.md [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+If you are only here for final version of the code, check out the [final branch](https://github.com/HarishTeens/rps-online/tree/final) of the same Repository.
+
+Here goes the file wise description of the Starter code,
+
+- ```public/index.html``` : A Home Page for the app with all the required UI Components. 
+- ```public/style.css``` : Styles for the entire application.
+- ```public/game.js``` : Client-Side JavaScript file to handle events.
+- ```app.js``` : Server Side NodeJS file for setting up the server.
+
+Note that the app is not functional yet, we need to write the necessary logic to make the app functional. Also, feel free to tweak the boiler code to match your style.
```
[the remark this answers is not in the document]
Comment (HarishTeens) [reply articles/rock-paper-scissor-online/index.md:34]: This sounds way better! Thanks

File: articles/rock-paper-scissor-online/index.md (hunk 1 above)
Comment (HarishTeens) [reply articles/rock-paper-scissor-online/index.md:36]: It felt a little better to have the file name highlighted though 😅  @louisefindlay23

**Comment by HarishTeens** [comment]
I was using Hemingway Editor to review my article. It flagged most of the sentences as hard to read. Hence I broke it down without using linking words. Now I realise, I should have trusted my human instincts more :/

File: articles/rock-paper-scissor-online/index.md [hunk 14]
```diff
@@ -0,0 +1,260 @@
+A variety of applications use Socket on the modern web. For example chat applications, real-time databases and online multiplayer games. All use Socket under the hood.
```
[the remark this answers is not in the document]
Comment (HarishTeens) [reply articles/rock-paper-scissor-online/index.md]: @louisefindlay23 I have edited the first paragraph and the line in prerequisites. Hope it clear things out.

**Comment by HarishTeens** [comment]
Hey, @louisefindlay23 
I have done the necessary suggested edits on the first half of the article, as well the second half based on your suggestions. Please check them out.

File: articles/rock-paper-scissor-online/index.md (hunk 6 above)
Comment (louisefindlay23) [reply articles/rock-paper-scissor-online/index.md]: Still need to add the comma after Next.

File: articles/rock-paper-scissor-online/index.md [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+    $("#message").html("Waiting for player 2, room ID is "+data.roomID).show();
+    roomID=data.roomID;
+})
+```
+
+The above code is added to `public/game.js`. This snippet lets player1 know that a room was created and player2 could use this roomID to join the room. Then, it hides the Room elements and displays a message with Room ID.
+
+### Join Game
+
+Next we write client-side logic on `public/game.js` to emit the `joinGame` event.
```
Comment (louisefindlay23) [inline articles/rock-paper-scissor-online/index.md:100]: Same here with next comma. Fix other instances.

File: articles/rock-paper-scissor-online/index.md [hunk 16]
```diff
@@ -0,0 +1,248 @@
...
+Next we write server-side logic on `app.js` to listen to the `joinGame` event and share each other info with the players.
+```
+//Join Game Listener
+    socket.on("joinGame",(data)=>{        
+        socket.join(data.roomID);
+        socket.to(data.roomID).emit("player2Joined",{p2name: data.name,p1name:players[data.roomID]});
+        socket.emit("player1Joined",{p2name:players[data.roomID],p1name:data.name});
+    })
+```
+First the above code adds player2 to the room, next it notifies player2 with player1's info and player1 with player2's info.
```
Comment (louisefindlay23) [inline articles/rock-paper-scissor-online/index.md:123]: Same comma situation with first.

File: articles/rock-paper-scissor-online/index.md [hunk 17]
```diff
@@ -0,0 +1,248 @@
...
+const result=(roomID)=> {
+    var winner = getWinner(choice1, choice2);
+    io.sockets.to(roomID).emit("result", {
+        winner: winner
+    });
+    choice1 = "";
+    choice2 = "";
+}
+```
+The result function takes the roomID as an argument. First, the `getWinner()` function calculates the result based on the player's choices .   Next, it emits the result to all the clients in the room, i.e both player1 and player2. Finally, it resets the player's choices.
```
Comment (louisefindlay23) [inline articles/rock-paper-scissor-online/index.md:207]: Extra spaces should be deleted before Next

File: articles/rock-paper-scissor-online/index.md [hunk 18]
```diff
@@ -0,0 +1,248 @@
...
+}
+```
+
+The above code listens to the result event and updates the DOM with the help of `updateDOM()` function. The `updateDOM()` function announces the winner and increases the score of the winner. In case of a draw, we just display a message that it's a draw.
+
+### Conclusion
+
+Kudos, You made it 🎉
+
+This app although works, yet could use a lot of improvement. As mentioned before, the intention was to keep it simple to put more focus Socket logic. Learning never stops here, now that the app works the next step is to do improvements in order to have a fruitful experience.
```
Comment (louisefindlay23) [inline articles/rock-paper-scissor-online/index.md:240]: **Flow Edit:** Although the app works, it 

add **on** after focus

I'd rephrase the entire learning never stops sentence. Something like: The next step is to improve the app by adding extra functionality. Some suggestions are **(suggest some additions to readers)**

**Review by louisefindlay23 — CHANGES_REQUESTED** [review]
Great work, @HarishTeens. The changes have improved the article a lot and thanks for taking my suggestions forward into the rest of the article. 

There's just a slight comma issue you should address and I've suggested some improvements to your conclusion.

File: articles/rock-paper-scissor-online/index.md (hunk 17 above)
Comment (HarishTeens) [reply articles/rock-paper-scissor-online/index.md:207]: removed all extra spaces

File: articles/rock-paper-scissor-online/index.md (hunk 18 above)
Comment (HarishTeens) [reply articles/rock-paper-scissor-online/index.md:240]: The new structure looks like,

...
The next step is to improve the app by adding extra functionality, read along for some suggestions.

### Next Steps
...

**Comment by HarishTeens** [comment]
Hey @louisefindlay23 , I've fixed all the final suggestions that you had mentioned as well. Hope this resolves everything.

**Review by louisefindlay23 — APPROVED** [review]
Congrats, @HarishTeens. I've made some small minor changes and your article is now ready for tech review.

**Review by louisefindlay23 — CHANGES_REQUESTED** [review]
@HarishTeens, I had an issue joining the game with a second player (see comments.) It would beneficial to add a complete code snippet for app.js and game.js at the Kudos section.

File: articles/rock-paper-scissor-online/index.md [hunk 19]
```diff
@@ -0,0 +1,255 @@
...
+    const winnerName=$("."+player+" .name").html().trim();
+    $("#message").html(winnerName+" scored a point!");
+}
+```
+
+The above code listens to the result event and updates the DOM with the help of `updateDOM()` function. The `updateDOM()` function announces the winner and increases the score of the winner. In case of a draw, we just display a message that it's a draw.
+
+### Conclusion
+
+Kudos, You made it 🎉
```
[the remark this answers is not in the document]
Comment (HarishTeens) [reply articles/rock-paper-scissor-online/index.md:243]: Sorry about that. it should have been `npm start`

File: articles/rock-paper-scissor-online/index.md [hunk 20]
```diff
@@ -0,0 +1,255 @@
...
+
+Although the app works, it could be improved. As mentioned before, the intention was to keep this tutorial simple to put more focus on Socket logic. The next step is to improve the app by adding extra functionality, read along for some suggestions.
```
[the remark this answers is not in the document]
Comment (HarishTeens) [reply articles/rock-paper-scissor-online/index.md:245]: Try a Hard Refresh before clicking on the Join button. One more possible reason could be, you might have put the socket events outside the `io.on` block. 

Yes, the final branch in the GitHub repo is the final version of the code, which I had attached at the beginning of the article.

**Comment by HarishTeens** [comment]
Hey @louisefindlay23 , 
I understand that it could be confusing to copy paste from the snippets of the article. I had given a thought about it and this was the best way to go about it, because If I were to write the logic of one file completely and then move on to the next file, the flow would have got hampered. I felt that not having too much code altogether could improve the reading experience, that is why I didn't add the final copy of the code in the bottom. 
One of the reason I had attached the final code at the beginning of the article is to avoid this confusion, if someone is just looking out for code. Hope you understand.

I had specified the file name on top of all snippets. Added the final version of `public/game.js` and `app.js` in the kudos section as you had suggested.

**Review by louisefindlay23 — APPROVED** [review]
@HarishTeens, thanks for making the changes. I've made some minor edits and your article is now ready for final review. @hectorkambow, can you add some front matter and publish?
